### PR TITLE
Add HashProtocol

### LINKED ISSUE #21: Add hash protocol
Also needs some sort of query key support.

### PR DESCRIPTION
**Pull request by taion**
Closes #21

### CODE REVIEW

File: src/HashProtocol.js [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+export default class HashProtocol {
+  constructor() {
+    this.stateStorage = new StateStorage(this, '@@farce');
+
+    this._index = null;
+    this._numExpectedHashChanges = 0;
+  }
+
+  init() {
+    // TODO: Do we still need to work around the old Firefox bug here?
```
Comment (taion) [inline src/HashProtocol.js]: This was a well-known Firefox bug, but I don't see it any more. Is it worth keeping the old workaround?

**Comment by jquense** [comment]
Is it update my OSS projects week?  :P

**Comment by taion** [comment]
It's itch-scratching week!

**Comment by taion** [comment]
I’m not going to bother for now. It was a weird feature anyway, and for most use cases, using the full path as key is sufficient.

**Comment by jquense** [comment]
fine by me, the use cases for this Protocol are increasingly minimal anyway

**Comment by taion** [comment]
It’s really just for TodoMVC ;)

**Pull request merged**